customization/bind-dirs.md

```diff
@@ -45,16 +52,32 @@ bind-dirs.sh works with Qubes R3.2 and above.
 * `/etc/qubes-bind-dirs.d`  (intermediate priority, for template wide configuration)
 * `/rw/config/qubes-bind-dirs.d` (highest priority, for per VM configuration)
 
+## How does it work? ##
+
+bind-dirs.sh is called on startup of a TemplateBasedVM, and configuration files in the configuration folders above are parsed to build a bash array.
+Files or folders identified in the array are copied to /rw/bind-dirs if they do not already exist there, and are then bind mounted over the original files/folders.
+
+Creation of the file and folders in /rw/bind-dirs should be automatic the first time the TemplateBasedVM is restarted after configuration.
+
+If you want to circumvent this process, you can create the relevant filestructure under /rw/bind-dirs and make any changes at the same time that you perform the configuration, before reboot.
+
+
 ## Limitations ##
 
-* Files that exist in the TemplateVM root image cannot be made deleted in the TemplateBasedVMs root image using bind-dirs.sh.
-* Does not work if the file / folder in question does not already exist in the root image. I.e. a file that does not exist in the root image cannot be bind mounted in the TemplateBasedVM.
-* Re-running `sudo /usr/lib/qubes/bind-dirs.sh` without previous `sudo /usr/lib/qubes/bind-dirs.sh umount` does not work.
+* Files that exist in the TemplateVM root image cannot be deleted in the TemplateBasedVMs root image using bind-dirs.sh.
+* The file / folder in question must already exist in the root image. I.e. a file that does not exist in the root image cannot be bind mounted in the TemplateBasedVM.
+* Re-running `sudo /usr/lib/qubes/bind-dirs.sh` without a previous `sudo /usr/lib/qubes/bind-dirs.sh umount` does not work.
 * Running 'sudo /usr/lib/qubes/bind-dirs.sh umount' after boot (before shutdown) is probably not sane and nothing can be done about that.
+* Many editors create a temporary file and copy it over the original file. If you have bind mounted an individual file this will break the mount.
```

Comment:
> copy it over the original file. If you have bind mounted an individual file this will break the mount.

This is not the actual limitation as I observed it (with `/etc/default/dnsmasq` being a bind mounted file by bind-dirs):

```
root@s-cm:/etc/default# cp dnsmasq dnsmasq-copy
root@s-cm:/etc/default# cp dnsmasq-copy dnsmasq

root@s-cm:/etc/default# strace -e trace=file mv dnsmasq-copy dnsmasq
[...]
rename("dnsmasq-copy", "dnsmasq")       = -1 EBUSY (Device or resource busy)
mv: cannot move ‘dnsmasq-copy’ to ‘dnsmasq’open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
: Device or resource busy
+++ exited with 1 +++

root@s-cm:/etc/default# strace -e trace=file rm dnsmasq
[...]
unlinkat(AT_FDCWD, "dnsmasq", 0)        = -1 EBUSY (Device or resource busy)
rm: cannot remove ‘dnsmasq’open("/usr/share/locale/en_US.UTF-8/LC_MESSAGES/libc.mo", O_RDONLY) = -1 ENOENT (No such file or directory)
: Device or resource busy
+++ exited with 1 +++
```

Moving/recreating is also a thing that config management systems like to do after templating a file (and validating it) which is also effected by this:

```
ansible-playbook -l s-cm ansible/ypid-playbooks/service/dnsmasq-qubes_os.yml
[...]
TASK [debops.dnsmasq : Configure defaults for dnsmasq service] *****************
fatal: [s-cm]: FAILED! => {"changed": true, "failed": true, "msg": "Could not replace file: /home/user/.ansible/tmp/ansible-tmp-1489657252.52-223355023089868/source to /etc/default/dnsmasq:
[Errno 16] Device or resource busy"}
```

Maybe generalize the point to: Deleting/renaming of bind mounted files/directories does not work. Programs might do this internally when updating a file by first creating a new file and them doing a `rename` or `unlinkat`. Examples include editors and configuration management systems. A workaround might be to bind mount the directory instead or do the file operations in the rw location (`/rw/bind-dirs/`).